src/Sarif.Multitool/Rules/SARIF1020.ReferToFinalSchema.cs

```diff
@@ -0,0 +1,55 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.Json.Pointer;
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.CodeAnalysis.Sarif.Multitool.Rules
+{
+    public class ReferToFinalSchema : SarifValidationSkimmerBase
+    {
+        private readonly MultiformatMessageString _fullDescription = new MultiformatMessageString
+        {
+            Text = RuleResources.SARIF1020_ReferToFinalSchema
+        };
+
+        public override MultiformatMessageString FullDescription => _fullDescription;
+
+        public override FailureLevel DefaultLevel => FailureLevel.Error;
+
+        /// <summary>
+        /// SARIF1020
+        /// </summary>
+        public override string Id => RuleId.ReferToFinalSchema;
+
+        protected override IEnumerable<string> MessageResourceNames => new string[]
+        {
+            nameof(RuleResources.SARIF1020_ReferenceToOldSchemaVersion),
+            nameof(RuleResources.SARIF1020_SchemaReferenceMissing)
+        };
+
+        protected override void Analyze(SarifLog log, string logPointer)
+        {
+            AnalyzeSchema(log.SchemaUri, logPointer);
+        }
+
+        private void AnalyzeSchema(Uri schemaUri, string pointer)
+        {
+            if (schemaUri == null)
+            {
+                LogResult(pointer.AtProperty(SarifPropertyName.Schema), nameof(RuleResources.SARIF1020_SchemaReferenceMissing));
```

Comment:
>pointer.AtProperty(SarifPropertyName.Schema) [](start = 26, length = 44)

As I mentioned earlier, in this case the `$schema` property _does not exist_. So we cannot specify the location of the `$schema` property (`/$schema`) as the location of the error.

Instead, we must specify the location of the root object (`/`) as the location of the error. That is just `pointer`. So the first argument to this error is `pointer`, not `pointer.AtProperty`.

Reply:
Think of it this way:

"The root object has a problem."
"What is the problem?"
"It doesn't have a `$schema` property"
"Oh! please tell me where the root object is so I can give it a `$schema` property."
"Here is it: `/`"

---
In reply to: [434101124](https://github.com/microsoft/sarif-sdk/pull/1892#discussion_r434101124) [](ancestors = 434101124)